compiler/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -116,13 +116,36 @@ struct PerFunctionContext<'f> {
 
     /// Track a value's last load across all blocks.
     /// If a value is not used in anymore loads we can remove the last store to that value.
-    last_loads: HashMap<ValueId, (InstructionId, BasicBlockId)>,
+    last_loads: HashMap<ValueId, (InstructionId, BasicBlockId, u32)>,
+
+    /// Track whether a load result was used across all blocks.
+    load_results: HashMap<ValueId, PerFuncLoadResultContext>,
+
+    /// Track whether a reference was passed into another entry point
+    /// This is needed to determine whether we can remove a store.
+    calls_reference_input: HashSet<ValueId>,
```

Comment:
Shouldn't passing a reference to a function already mark the reference's value as unknown? What is the second hashset needed for?

Reply:
last stores don't exist in SSA generally so I'm assuming you mean the allocations themselves. Those should already be handled by existing checks

Reply:
Yeah so the existing check does not remove these stores. I added this map as when I am doing a function cleanup we may spot this lone store without any loads and remove it as we never load from this address. However, we cannot remove it due to this reference arg case.

Reply:
What if the called function mutates the address to alias it with another?
Would probably require double references like `let mut x = &mut 0` or `let y = &mut &mut 0`.